index.js

```diff
@@ -2,32 +2,34 @@ import Punycode from 'punycode/punycode.js';
 
 import rules from './data/rules.js';
 
-
 var internals = {};
 
 
 //
 // Parse rules from file.
 //
-internals.rules = rules.map(function (rule) {
-
-  return {
-    rule: rule,
-    suffix: rule.replace(/^(\*\.|\!)/, ''),
-    punySuffix: -1,
-    wildcard: rule.charAt(0) === '*',
-    exception: rule.charAt(0) === '!'
-  };
-});
+const rulesByPunySuffix = rules.reduce(
+  (map, rule) => {
+    const suffix = rule.replace(/^(\*\.|\!)/, '');
+    const punySuffix = Punycode.toASCII(suffix);
+    const firstChar = rule.charAt(0);
+
+    if (map.has(punySuffix)) {
+      throw new Error(`Multiple rules found for ${rule} (${punySuffix})`);
+    }
 
+    map.set(punySuffix, {
+      rule,
+      suffix,
+      punySuffix,
+      wildcard: firstChar === '*',
+      exception: firstChar === '!'
+    });
 
-//
-// Check if given string ends with `suffix`.
-//
-internals.endsWith = function (str, suffix) {
-
-  return str.indexOf(suffix, str.length - suffix.length) !== -1;
-};
+    return map;
+  },
+  new Map(),
+);
```

Comment:
Brilliant.

I tried to find a case/scenario were it could "break", I tried to check with incredibly long URLs like:

```
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb.cccccccccccccccccccccccccccccccccccccccccccccccccccccc.dddddddddddddddddddddddddddddddddddddddddddddddddddddd.com
```

But even then I still got `'64.70x up'` on the benchmark run.

I also tried `very.complex.nested.subdomain.example.co.uk` and it also works like a charm.

I feel like I've got nothing of substance to add, this just works great, you've reduced the complexity significantly. :clap: :clap:

Reply:
Having you test it locally and check the code gives me confidence to merge this :muscle: